packages/payment-processor/src/payment/single-request-proxy.ts

```diff
@@ -0,0 +1,188 @@
+      (isERC20 ? 'ERC20SingleRequestProxyCreated' : 'EthereumSingleRequestProxyCreated'),
+  );
+
+  if (!event) {
+    throw new Error('Single request proxy creation event not found');
+  }
+
+  const proxyAddress = event.args?.[0];
+  if (!proxyAddress) {
+    throw new Error('Proxy address not found in event args');
+  }
+
+  return proxyAddress;
+}
+
+/**
+ * Executes a payment through a Single Request Proxy contract.
+ *
+ * @param singleRequestProxyAddress - The address of the deployed Single Request Proxy contract.
+ * @param signer - The Ethereum signer used to execute the payment transaction.
+ * @param amount - The amount to be paid, as a string representation of the value.
+ * @returns A Promise that resolves when the payment transaction is confirmed.
+ * @throws {Error} If the SingleRequestProxy contract is invalid.
+ * @throws {Error} If the proxy contract type cannot be determined, or if any transaction fails.
+ *
+ * @remarks
+ * This function supports both ERC20 and Ethereum payments.
+ * For ERC20 payments, it first transfers the tokens to the proxy contract and then triggers the payment.
+ * For Ethereum payments, it directly sends the Ether to the proxy contract.
+ * The function automatically detects whether the proxy is for ERC20 or Ethereum based on the contract interface.
+ */
+export async function payRequestWithSingleRequestProxy(
+  singleRequestProxyAddress: string,
+  signer: Signer,
+  amount: string,
+): Promise<void> {
+  // Create contract interface with all required methods to validate SingleRequestProxy
+  const proxyContract = new Contract(
+    singleRequestProxyAddress,
+    [
+      'function payee() view returns (address)',
+      'function paymentReference() view returns (bytes)',
+      'function feeAddress() view returns (address)',
+      'function feeAmount() view returns (uint256)',
+      'function tokenAddress() view returns (address)',
+    ],
+    signer,
+  );
+
+  // Validate that this is a SingleRequestProxy by checking required methods
+  try {
+    await Promise.all([
+      proxyContract.payee(),
+      proxyContract.paymentReference(),
+      proxyContract.feeAddress(),
+      proxyContract.feeAmount(),
+    ]);
+  } catch (error) {
+    throw new Error('Invalid SingleRequestProxy contract');
+  }
+
+  let isERC20: boolean;
+  let tokenAddress: string | null = null;
+  try {
+    tokenAddress = await proxyContract.tokenAddress();
+    isERC20 = true;
+  } catch {
+    isERC20 = false;
+  }
+
+  if (isERC20 && tokenAddress) {
+    // ERC20 payment
+    const erc20Contract = IERC20__factory.connect(tokenAddress, signer);
+
+    // Transfer tokens to the proxy
+    const transferTx = await erc20Contract.transfer(singleRequestProxyAddress, amount);
+    await transferTx.wait();
+
+    // Trigger the receive function with 0 ETH
+    const triggerTx = await signer.sendTransaction({
+      to: singleRequestProxyAddress,
+      value: ethers.constants.Zero,
+    });
+    await triggerTx.wait();
```

Comment:
Is it possible to send ERC20 to the SingleRequestProxy and then send a 0 ETH call, in the same transaction - via a proxy contract?  Like an "ERC20SingleRequestProxyProxy" for lack of a better name.

Reply:
We discussed this at length at the Tech Daily today.  Conclusion: It may be possible, but we don't want to delay the feature. We should investigate this optimization later.

Reply:
We realized that optimizing to allow sending ERC20 and 0 ETH call in the same transaction requires a function call 😵‍💫 but the whole point of the SingleRequestProxy is to eliminate the need for function calls. This further confirms that we should hold off on optimizing for now - but we feel more confident that we _could_ probably build it later if someone asks for it.